Issue creation

sebastinez opened the pull request:
## Pending issues
- [x] The issue listing is being paginated with 10 issues per page, which impacts the issue count shown of # open and closed.
    This PR should improve this https://github.com/radicle-dev/heartwood/pull/355
- [x] The design is being worked on by brandon and I'm making still changes.
- [x] There is stuff to be refactored, due to changes in the format returned by the httpd with recent changes in Dids and Authors.
- [x] Add validation captions to assignees
- [x] The modal shown if an issue can't be created, needs a better explainer for end users.
- [x] Once the design is more fixed, I want to add some e2e tests which go through the creation flow
    Will keep this functionality out of this PR, since it requires a bit more work, than I want to put into this PR.
- [x] Add some redirection or simil to notify users why they aren't able to create issues when logged out.
- [x] Add responsiveness to creation form

Adds creation of issues from a logged in web interface.

We get a new `New issue` button on the top right side of the issue listing.
The button is disabled if no session is found.
<img width="670" alt="image" src="https://user-images.githubusercontent.com/7912302/219472667-b105209a-edc8-456d-b517-37e8cde9fd40.png">

On the issue creation screen we have one textfield for the title and description,
and on the right side a textfield for assignees and labels to provide for issue creation.
Also there is a `Preview` button that renders everything into Markdown.
![image](https://user-images.githubusercontent.com/7912302/220181057-e20b5fd6-6dde-4e0f-91a9-ec3e2b4bab22.png)

Here a view of a completed issue creation form
![image](https://user-images.githubusercontent.com/7912302/220181216-ec3570e0-cb82-4dc5-a1c6-ea4f7595101e.png)

Once clicked the issue gets created and on success the user gets redirected to an updated issue listing.
Any issue during the issue creation, will show a FailureModal that eventually will explain why the issue hasn't been created.

cloudhead commented:
Nice work!

Some feedback on the ux:
* We should split the title from the issue body, since users have to specify both anyway
* On success, we should just show the issue, no need for a success modal
* The preview button should probably be somewhere else, and not be a button, but we can explore that further

cloudhead commented:
You could also maybe try re-using the styles of issue/patch comments for the preview :thinking:

rudolfs reviewed (commented):
This is great! Looking forward to starting using it instead of GH :) 🎉 💯 

Here are my thoughts on the design/UX, perhaps @cloudhead can first chip in whether he agrees before making any changes to avoid unnecessary churn.

- We should avoid highlighting affecting positioning. To avoid this we can just set a border in the unhighlighted state that is the same color as the background:

  https://user-images.githubusercontent.com/158411/220601926-6f37e20c-162d-472f-9a5c-62dfe0110d91.mov

- Can we have vegetarian empty states? : D Or just use the standard one 🍂 that we use elsewhere?

  <img width="372" alt="Screenshot 2023-02-22 at 10 50 07" src="https://user-images.githubusercontent.com/158411/220602423-ca17ab54-1e47-4038-9766-7b564440e997.png">
  <img width="375" alt="Screenshot 2023-02-22 at 10 50 09" src="https://user-images.githubusercontent.com/158411/220602424-ce0ee355-a915-471a-9e87-b4f0fa612ddc.png">

- Some wording suggestions:
  <img width="235" alt="Screenshot 2023-02-22 at 10 53 16" src="https://user-images.githubusercontent.com/158411/220602616-0ee762c6-6bd6-4856-b65f-d2b9c6816c87.png">
  - "No asignees" (plural, without a dot)
  - "No tags"
  <img width="240" alt="Screenshot 2023-02-22 at 11 12 03" src="https://user-images.githubusercontent.com/158411/220605344-926240b0-0d90-48db-866d-b0e857cb11bb.png">
  - "This tag already exists"
  <img width="534" alt="Screenshot 2023-02-22 at 11 13 52" src="https://user-images.githubusercontent.com/158411/220605444-98c295be-2006-4718-9334-1b0c67f9c850.png">
  - "This user is already assigned"
- The spacing between tags and users feels a bit too sparse, maybe reduce it to 0.5rem?
  <img width="184" alt="Screenshot 2023-02-22 at 11 14 25" src="https://user-images.githubusercontent.com/158411/220605624-97e16177-038c-4b1d-9da8-c6b5b1fe831c.png">

- The spacing around the buttons could be bigger, like 1rem and there should be a top margin as well. Maybe the preview button can be of the `text` variant?

<img width="586" alt="Screenshot 2023-02-22 at 12 22 36" src="https://user-images.githubusercontent.com/158411/220606399-3d9ef2d4-5f30-4297-9a17-869387f216b5.png">


- Could we use the comment icon from the [new icon set](https://github.com/radicle-dev/radicle-interface/issues/604) instead of an emoji?
  <img width="671" alt="Screenshot 2023-02-22 at 12 11 08" src="https://user-images.githubusercontent.com/158411/220604132-635fe21c-5e91-44f2-8823-5a1e5c340c78.png">
  <img width="44" alt="Screenshot 2023-02-22 at 10 54 20" src="https://user-images.githubusercontent.com/158411/220604310-b836453e-473b-49b6-b67c-172b700a6c8e.png">

- I think there should be some indication that we're in preview mode. The `Preview` button could maybe be called `Resume editing` perhaps?

- Should we use the default error 🚨 emoji for the modal? The wording could also be improved to avoid the word "issue" twice. How about "Could not create the issue. Make sure you're still logged in."?

  <img width="711" alt="Screenshot 2023-02-22 at 11 15 15" src="https://user-images.githubusercontent.com/158411/220607012-4b3134a6-30bb-400b-8b1b-b522973d1c44.png">

- Can we remove the context menu? It doesn't seem functional at the moment.
  <img width="86" alt="Screenshot 2023-02-22 at 12 30 15" src="https://user-images.githubusercontent.com/158411/220607848-fe658ba1-4930-4fdc-ba18-dac8fb23e33c.png">

- I'd also hide the patches tab for now if there's no functionality for now.
<img width="1120" alt="Screenshot 2023-02-22 at 12 31 27" src="https://user-images.githubusercontent.com/158411/220608032-b45d852d-e9fc-4f1a-9eaf-4edfb94b0400.png">

- Wonder if we should maybe keep the input box always under the heading, so it stays at the top if there are a lot of assignees/tags? Also the wording could be "Assign issue to, ex: did:key:z6M…" and "Tag issue".
  <img width="583" alt="Screenshot 2023-02-22 at 12 09 55" src="https://user-images.githubusercontent.com/158411/220609534-6184e7b1-82c5-4ea2-956d-f7438be41a11.png">

------

And here are some functional problems I found:

- this happened after I added a couple of issues, one of them had a markdown title in the body:
  <img width="1840" alt="Screenshot 2023-02-22 at 10 58 55" src="https://user-images.githubusercontent.com/158411/220608428-eb037f0a-8dd7-4dc9-98d2-5c8f7126023a.png">

- the assignee validation seems buggy, if I enter a valid did, the form becomes valid and when I delete a couple of characters from the did, it is still valid and i can submit it with the invalid did:
  <img width="530" alt="Screenshot 2023-02-22 at 11 13 35" src="https://user-images.githubusercontent.com/158411/220608720-15997ae3-7d8f-4b02-b9ae-aeb2f526156b.png">

sebastinez commented:
> * We should avoid highlighting affecting positioning. To avoid this we can just set a border in the unhighlighted state that is the same color as the background

This has been implemented

> * Can we have vegetarian empty states? : D Or just use the standard one 🍂 that we use elsewhere?

I added the standard one in both places

> * Some wording suggestions:

I added the wording suggestions

Fixed the spacing and changed the preview button to the `text` variant

> * Could we use the comment icon from the [new icon set](https://github.com/radicle-dev/radicle-interface/issues/604) instead of an emoji?

Done

> * I think there should be some indication that we're in preview mode. The `Preview` button could maybe be called `Resume editing` perhaps?

Done

> * Should we use the default error 🚨 emoji for the modal? The wording could also be improved to avoid the word "issue" twice. How about "Could not create the issue. Make sure you're still logged in."?

Done

> * Can we remove the context menu? It doesn't seem functional at the moment.

Done

> * I'd also hide the patches tab for now if there's no functionality for now.

I want to keep the patches tab for now, since it has read-only functionality, and in a short time it'll be functional.

> * Wonder if we should maybe keep the input box always under the heading, so it stays at the top if there are a lot of assignees/tags? Also the wording could be "Assign issue to, ex: did🔑z6M…" and "Tag issue".

I tried that out, but I didn't like to have an empty input field stay in between the title and the inputs, IMO the best way would be to have the inputs as "pills" inside of the `TextInput` but maybe we find a better way in the future.

> And here are some functional problems I found:
> 
> * this happened after I added a couple of issues, one of them had a markdown title in the body:

It seems that there is some serialization issue, which I'm trying to fix currently

> * the assignee validation seems buggy, if I enter a valid did, the form becomes valid and when I delete a couple of characters from the did, it is still valid and i can submit it with the invalid did:

This should be better now

> * when I restarted the backend, the UI still shows me as "logged in" (i.e. the connect button still shows my identity), but when I try to create an issue, I get the error modal

This has also been improved, but one has to wait 60 seconds for the web ui to realize that the backend has been shutdown or restarted (maybe we should reduce it to less time :thinking:)

sebastinez commented:
Ok this issue is important for this PR to make sense, since else we aren't able to handle multiline issues descriptions, or comments for that matter.

https://github.com/radicle-dev/heartwood/issues/349

So I'll put this PR on hold until that issue has been closed.

sebastinez commented on src/lib/utils.ts:
```diff
@@ -63,6 +64,13 @@ export function parseNid(
+    const hex = bs58.decode(match[2].substring(1));
+    // This checks also that the first 2 bytes are equal
+    // to the ed25519 public key type used.
+    if (!(hex.byteLength === 34 && hex[0] === 237 && hex[1] === 1)) {
```
Thanks @cloudhead have added this in 2877f28

rudolfs commented on src/views/projects/Issue/New.svelte:
```diff
@@ -0,0 +1,283 @@
+    <div class="editor">
+      {#if preview}
+        <div class="summary txt-medium" class:txt-missing={preview}>
+          {issueTitle || "No title"}
```
```suggestion
          {issueTitle ?? "No title"}
```

sebastinez replied:
In the end I changed this to be more similar to how we handle it in `Comment` to trim the `issueTitle` and check strict equal against an empty string

rudolfs reviewed (commented):
Nice work!! 🎉 

Added some suggestions.

sebastinez commented on src/lib/router.ts:
```diff
@@ -344,10 +355,23 @@ function resolveProjectRoute(
-    if (issue) {
+    const issueOrAction = segments.shift();
+    if (issueOrAction === "new") {
+      if (!session) {
+        pop(); // Go back if no session is available.
+      }
```
Yeah you're right, the reason for this check is if you visit `http://localhost:3000/seeds/0.0.0.0/rad:z2H9aHDurxd8Uvx2jsvW4e5mamy5S/issues/new` without having a session in place has to be checked somehow, and one way I thought of is to just send the user back.
Another way was to show a error modal but I didn't like the approach.
